src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.cs

```diff
@@ -32,68 +33,93 @@ public partial class EventSourceGenerator : ISourceGenerator
         //         }
         //     }
 
-        public void Initialize(GeneratorInitializationContext context)
-            => context.RegisterForSyntaxNotifications(() => new SyntaxReceiver());
+        public void Initialize(IncrementalGeneratorInitializationContext context)
+        {
+            IncrementalValuesProvider<EventSourceClass> eventSourceClasses =
+                context.SyntaxProvider
+                .CreateSyntaxProvider(static (x, _) => IsSyntaxTargetForGeneration(x), (x, _) => GetSemanticTargetForGeneration(x))
+                .Where(x => x is not null)
+                .Collect()
+                .Combine(context.CompilationProvider)
+                // Use a custom comparer that ignores the compilation.
+                .WithComparer(ClassDeclarationsEqualityComparer.Instance)
```

Comment:
@chsienki has encouraged us in the past to do this, and we do in our other generators. Why is this case different?

Reply:
This case is not different. The RegexGenerator is also broken (e.g. if you start with a project that allows unsafe, and then you alter the project where the _only change_ is to disable unsafe code, RegexGenerator will fail to detect the change and will keep the original code).

Reply:
IIRC, @chsienki recommended doing this because otherwise any change anywhere in the project (or maybe file) would invalidate caching, and thus basically caching was useless.  Is that not the case?

Reply:
The correct approach for RegexGenerator is to extract a subset of information from the compilation into a comparable, self-contained data model (e.g. one property it would have is `AllowsUnsafe`). Here's an example from another generator:

https://github.com/microsoft/vs-extension-testing/blob/6e2208049bcc1581883cd7d146a2f37e490b9fb0/src/Microsoft.VisualStudio.Extensibility.Testing.SourceGenerator/TestServicesSourceGenerator.cs#L617-L632

In practice, this data extract step will run fairly often, but it will generally produce the same result as the previous incremental generation and thus not trigger downstream steps to run.

Reply:
Thanks @sharwell, I will keep that in mind.

And I am interested to fix the other generators accordingly.